test/active_job/queue_adapters/amazon_sqs_adapter_test.rb

```diff
@@ -29,20 +29,71 @@ module QueueAdapters
         end
 
         it 'adds message_deduplication_id and default message_group_id if job does not override it' do
-          expect(client).to receive(:send_message)
-                              .with(
-                                {
-                                  queue_url: 'https://queue-url.fifo',
-                                  message_body: instance_of(String),
-                                  message_attributes: instance_of(Hash),
-                                  message_group_id: Aws::Rails::SqsActiveJob.config.message_group_id,
-                                  message_deduplication_id: instance_of(String)
-                                }
-                              )
+          expect(client).to receive(:send_message).with(
+            {
+              queue_url: 'https://queue-url.fifo',
+              message_body: instance_of(String),
+              message_attributes: instance_of(Hash),
+              message_group_id: Aws::Rails::SqsActiveJob.config.message_group_id,
+              message_deduplication_id: instance_of(String)
+            }
+          )
+
           TestJob.perform_later('test')
           sleep(0.2)
         end
 
+        describe 'when job has excluded deduplication keys defined' do
+          describe 'through #deduplicate_without' do
+            let(:ex_dedup_keys) { ['job_class', 'queue_name'] }
+            let(:ex_dudup_keys_with_job_id) { ex_dedup_keys << 'job_id' }
+            before do
+              allow_any_instance_of(TestJobWithDedupKeys).to receive(:excluded_deduplication_keys)
+                .and_return(ex_dudup_keys_with_job_id)
+            end
+            it 'adds customized message_deduplication_id' do
+              expect_any_instance_of(TestJobWithDedupKeys).to receive(:excluded_deduplication_keys).once
+              expect(client).to receive(:send_message).with(
+                {
+                  queue_url: 'https://queue-url.fifo',
+                  message_body: instance_of(String),
+                  message_attributes: instance_of(Hash),
+                  message_group_id: Aws::Rails::SqsActiveJob.config.message_group_id,
+                  message_deduplication_id: instance_of(String)
```

Comment:
Is there a way to strongly assert that message_deduplication_id does not contain job_class and queue_name here? Maybe an implementation detail but we could assert `Digest::SHA256.hexdigest` or `Aws::Json.dump` doesn't receive them.